plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerContentProvider.java

```diff
@@ -56,16 +57,37 @@ protected void handleConnectionChanged(IConnection connection, String property,
 				refreshGrouping(groupMap.get(resource.getProject()), resource.getKind());
 			}else if(oldValue != null && newValue == null) {
 				//delete
-				IResource resource = (IResource)oldValue;
+				IResource resource = (IResource) oldValue;
 				refreshGrouping(groupMap.get(resource.getProject()), resource.getKind());
 			}else {
 				refreshViewer(newValue);
 			}
+		}else if(ConnectionProperties.PROPERTY_PROJECTS.equals(property)){
+			handleProjectChanges(oldValue, newValue);
 		}else{
 			super.handleConnectionChanged(connection, property, oldValue, newValue);
 		}
 	}
 	
+	@SuppressWarnings("unchecked")
+	private void handleProjectChanges(Object oldValue, Object newValue) {
+		List<IProject> newProjects = (List<IProject>) newValue;
+		List<IProject> oldProjects = (List<IProject>) oldValue;
+		//theory this should only be one.
+		Object [] diffs = CollectionUtils.disjunction(newProjects, oldProjects).toArray();
+		if(newProjects.size() > oldProjects.size()) {
```

Comment:
is the case handled where new projects were added but a project was removed too (which is actually a case where children have to be removed AND children added)?